backend/src/composite/split.rs

```diff
@@ -134,54 +168,69 @@ fn split_by_namespace<F: FieldElement>(
         })
 }
 
+/// Merges namespaces without any polynomials into the other namespaces.
+/// For example, a hint might reference a symbol in an std namespace, so we just make
+/// those available to all machines.
+fn merge_empty_namespaces(
+    statements_by_namespace: BTreeMap<String, Vec<StatementIdentifier>>,
+    definitions: &HashMap<String, (Symbol, Option<FunctionValueDefinition>)>,
+) -> BTreeMap<String, Vec<StatementIdentifier>> {
+    // Separate out machines without any polynomials
+    let (proper_machines, empty_machines) = statements_by_namespace
+        .into_iter()
+        .partition::<BTreeMap<_, _>, _>(|(_, statements)| {
+            statements.iter().any(|statement| match statement {
+                StatementIdentifier::Definition(name) => {
+                    matches!(definitions[name].0.kind, SymbolKind::Poly(_))
+                }
+                _ => false,
+            })
+        });
+    // Merge empty machines into the proper machines
+    proper_machines
+        .into_iter()
+        .map(|(namespace, mut statements)| {
+            statements.extend(
+                empty_machines
+                    .iter()
+                    .flat_map(|(_, statements)| statements.clone()),
+            );
+            (namespace, statements)
+        })
+        .collect::<BTreeMap<_, _>>()
+}
+
 /// Given a PIL and a list of statements, returns a new PIL that only contains the
 /// given subset of statements.
-/// Returns None if there are no identities in the subset of statements.
 fn build_machine_pil<F: FieldElement>(
     pil: Analyzed<F>,
     statements: Vec<StatementIdentifier>,
-) -> Option<Analyzed<F>> {
-    // TODO: After #1488 is fixed, we can implement this like so:
-    // let pil = Analyzed {
-    //     source_order: statements,
-    //     ..pil.clone()
-    // };
-    // let parsed_string = powdr_parser::parse(None, &pil.to_string()).unwrap();
-    // let pil = powdr_pil_analyzer::analyze_ast(parsed_string);
+) -> Analyzed<F> {
+    let pil = Analyzed {
+        source_order: statements,
+        ..pil
+    };
+    let pil_string = add_dummy_witness_column(&pil.to_string());
+    let parsed_string = powdr_parser::parse(None, &pil_string).unwrap();
+    powdr_pil_analyzer::analyze_ast(parsed_string)
+}
 
-    // HACK: Replace unreferenced identities with 0 = 0, to avoid having to re-assign IDs.
-    let identities = statements
+/// Insert a dummy witness column and identity into the PIL string, just after the namespace.
```

Comment:
This does not explain why it's required to have a witness and identity? Is it because of the absence of proof object?

Reply:
I added a comment.

Reply:
I'm not convinced by this, I think it's something which should be done by each backend.

Reply:
IDK, I think it's fair that you can't create a SNARK without a witness or fixed column, because if there are only fixed columns, you could validated whatever property already when checking the verification key. And also, as mentioned, we'll never have this situation in practice as soon as we implement the Bus.

Reply:
I agree that it's an edge case that typically does not happen, but I don't understand why it needs to be fixed in this PR. My reasoning is that let's say you have an empty circuit that you send through the composite backend to halo2. Here it would make sure halo2 does not fail, by adding a dummy constraint. But then if you send the same empty circuit straight to halo2, halo2 would fail, so you'd have to fix halo2. At which point you don't need this change in the composite backend.

Reply:
Hmm, but I guess my argument is that it's OK for Halo2 to fail, because it doesn't make sense to prove a circuit without any witness column and / or constraint. It's the composite backend's "fault" for passing such a circuit, and it only currently does so because we haven't implemented all features. 

You could also take the view that the backend should handle all circuits, even those that don't make sense. I initially fixed it in Halo2, but it led to pretty ugly code and it seems plausible we'll have to do the same for other backends. I didn't really see the value of that code, so I pivoted to instead declaring these circuits as invalid and fixing the composite backend. Eventually, it won't be necessary there as well.

I think it's a fair approach, but I can also make sure the backends support empty circuits if you have a strong opinion on it.

Reply:
> I don't understand why it needs to be fixed in this PR.

The composite backend creates empty circuits by removing lookup / permutation constraints and proving the connected components separately. Because that creates empty circuits (e.g. for the RHS of fixed lookups, which only consists of fixed columns), it has to be addressed somehow.

Reply:
Oh ok I did not understand that it was because we're removing constraints.
In general I think all backends should behave the same with regards to empty circuits, whether that is failing or inserting dummy constraints themselves, but that's unrelated.